### ISSUE TITLE
Get "https://localhost:4000": x509: certificate is valid for localhost, not hostname

### ISSUE THREAD

**Issue by DejfCold** [hashicorp/consul#10501]
#### Overview of the Issue
After upgrading to consul-1.10.0-1.x86_64 (from 1.9.7 I think) my https service checks started to fail with `Get "https://localhost:4000": x509: certificate is valid for localhost, not hostname` where hostname is my actual hostname.

`curl https://localhost:4000` works. curl for hostname obviously doesn't, but it also shouldn't. I could probably work around it by setting `tls_skip_verify` to true, but I don't think that's a good option.

But it's weird. Why would it try to check the certificate against hostname when it clearly says it's trying to use localhost?

#### Reproduction Steps

1. Add the following to /etc/consul.d/service.hcl and update the token value
```hcl
service = {
  id = "hostname-service-4000"
  name = "service"
  port = 4000
  checks = [
    {
      tcp = "localhost:4000"
      name = "tcp check"
      interval = "10s"
      timeout = "1s"
    },
    {
      http = "https://localhost:4000"
      name = "Https check"
      tls_skip_verify = false
      interval = "10s"
      timeout = "1s"
    }
  ]
  token = "xxx"
}
```
2. generate self-signed certificate for `localhost`, `service.service.consul` and add it's public key to your public keychain, so consul trusts it and place both private and public key parts to /etc/service/
3. create the following nginx config:
```conf
server {
  listen 4000 ssl http2;
  listen [::]:4000 ssl http2;
  server_name service.service.consul _;
  ssl_certificate /etc/service/cert.pem;
  ssl_certificate_key /etc/service/key.pem;

  location / {
    root /usr/share/nginx/service;
    autoindex on;
  }
}
```
4. reload nginx
5. reload or restart consul

1. run `consul monitor` > `[WARN]  agent: Check is now critical: check=service:hostname-service-4000:2`

### Consul info for both Client and Server

```
consul info
agent:
        check_monitors = 0
        check_ttls = 2
        checks = 10
        services = 7
build:
        prerelease = 
        revision = 27de64da
        version = 1.10.0
consul:
        acl = enabled
        bootstrap = false
        known_datacenters = 1
        leader = false
        leader_addr = xxxx:8300
        server = true
raft:
        applied_index = 5043160
        commit_index = 5043160
        fsm_pending = 0
        last_contact = 16.959054ms
        last_log_index = 5043160
        last_log_term = 6477
        last_snapshot_index = 5033205
        last_snapshot_term = 6475
        latest_configuration = [{Suffrage:Voter ID:7e9de0e6-f79e-20b5-db7f-08c87502521b Address:zzzz:8300} {Suffrage:Voter ID:4a3ce90f-9ef4-9ca8-9d16-8ecd32986673 Address:xxxx:8300} {Suffrage:Voter ID:3c92dedc-5478-e20f-339c-e0e99652f226 Address:yyyy:8300}]
        latest_configuration_index = 0
        num_peers = 2
        protocol_version = 3
        protocol_version_max = 3
        protocol_version_min = 0
        snapshot_version_max = 1
        snapshot_version_min = 0
        state = Follower
        term = 6477
runtime:
        arch = amd64
        cpu_count = 4
        goroutines = 128
        max_procs = 4
        os = linux
        version = go1.16.5
serf_lan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 899
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 19682
        members = 3
        query_queue = 0
        query_time = 1
serf_wan:
        coordinate_resets = 0
        encrypted = true
        event_queue = 0
        event_time = 1
        failed = 0
        health_score = 0
        intent_queue = 0
        left = 0
        member_time = 5560
        members = 3
        query_queue = 0
        query_time = 1
```

</details>

### Operating system and Environment details
Centos 8.3 and Rocky Linux 8.4 VMs

### Log Fragments

![image](https://user-images.githubusercontent.com/12591578/123361524-81c21b00-d56f-11eb-9c34-6dc9be3c13cb.png)

**Comment by DejfCold** [comment]
Oh, right, I suppose this is kinda duplicate of #10490?

**Comment by blake** [comment]
@DejfCold Yup, this appears to be duplicate of that issue. A fix will be available in an upcoming 1.10.x patch release.
